### PR TITLE
Add tokens table

### PR DESCRIPTION
**Pull request by allouis**
This is a table to store single use tokens for use in magic links, the
columns are as simple as possible at the moment and are designed as:

`id` - standard ObjectID like all of our tables
`token` - 128bit base64 encoded string
`data` - arbitrary data to store against the token
`created_at` - timestamp to allow for expiry to be implemented for tokens

### CODE REVIEW

**Review by daniellockyer — CHANGES_REQUESTED** [review]

File: core/server/data/migrations/utils.js [hunk 1]
```diff
@@ -1,8 +1,37 @@
...
 
+/**
+ * Creates a migrations which will add a new table from schema.js to the database
+ */
+function addTable(name) {
+    return createNonTransactionalMigration(
+        async function up(connection) {
+            const tableExists = await connection.schema.hasTable(name);
+            if (tableExists) {
+                logging.warn(`Adding table: ${name}`);
```
[the remark this answers is not in the document]
Comment (allouis) [reply core/server/data/migrations/utils.js]: Dun!

File: core/server/data/migrations/utils.js [hunk 2]
```diff
@@ -1,8 +1,37 @@
...
+                return;
+            }
+
+            logging.info(`Adding table: ${name}`);
+            return commands.createTable(name, connection);
+        },
+        async function down(connection) {
+            const tableExists = await connection.schema.hasTable(name);
+            if (!tableExists) {
+                logging.warn(`Dropping table: ${name}`);
```
[the remark this answers is not in the document]
Comment (allouis) [reply core/server/data/migrations/utils.js]: And dun!

**Review by rshbhgrg — COMMENTED** [review]
Changes look good, just wondering if we should add any of the other standard columns like `created_by`, `updated_at`, `updated_by` here, mostly `created_by` and then hook it up with actions table which helps us keep internal track of who created a token etc for any checks later?  🤔

**Comment by kevinansfield** [comment]
> Changes look good, just wondering if we should add any of the other standard columns like `created_by`, `updated_at`, `updated_by` here, mostly `created_by` and then hook it up with actions table which helps us keep internal track of who created a token etc for any checks later? 🤔

Is there anything in this table that would be updated after creation?

**Review by kevinansfield — APPROVED** [review]
Looks good 👍 If we have a `created_at` we should probably have a `created_by` too if these resources can be created by authenticated users rather than as a system side-effect.

**Comment by allouis** [comment]
@kevinansfield 
> Is there anything in this table that would be updated after creation?

Nope! These are created, and then as soon as they are read for the first time, they will be deleted :relaxed:

> If we have a created_at we should probably have a created_by too if these resources can be created by authenticated users rather than as a system side-effect.

:+1: I'm gonna add that now!

**Review by rshbhgrg — APPROVED** [review]

**Review by daniellockyer — APPROVED** [review]

**Pull request merged**